src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Compiler/FeatureFlagPass.php

```diff
@@ -0,0 +1,92 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ChildDefinition;
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Reference;
+use Symfony\Component\FeatureFlag\Debug\TraceableFeatureChecker;
+
+class FeatureFlagPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container): void
+    {
+        if (!$container->hasDefinition('feature_flag.feature_checker')) {
+            return;
+        }
+
+        $features = [];
+        foreach ($container->findTaggedServiceIds('feature_flag.feature') as $serviceId => $tags) {
+            $className = $this->getServiceClass($container, $serviceId);
+            $r = $container->getReflectionClass($className);
+
+            if (null === $r) {
+                throw new \RuntimeException(\sprintf('Invalid service "%s": class "%s" does not exist.', $serviceId, $className));
+            }
+
+            foreach ($tags as $tag) {
+                $featureName = ($tag['feature'] ?? '') ?: $className;
+                if (\array_key_exists($featureName, $features)) {
+                    throw new \RuntimeException(\sprintf('Feature "%s" already defined.', $featureName));
```

Comment:
the message should be clear that it does allows multiple feature to have the same name but not in the same provider IMO

Reply:
For now, all features declared using a tag (or indirectly, the `#[AsFeature]` attribute) will be used by the `InMemoryProvider`. What about a message like `The "%s" feature is already defined in the InMemoryProvider` ?

Reply:
Replaced by `Feature "%s" already defined in the "feature_flag.provider.in_memory" provider.` since the service's name is hard-coded below.
```php
$container->getDefinition('feature_flag.provider.in_memory')
            ->setArgument('$features', $features)
        ;
```